### PR TITLE
Created a new feature/dma2 branch from the master branch

### PR DESCRIPTION
**Pull request by lzhan55**

### CODE REVIEW

File: libopae/plugins/xfpga/feature_plugins/feature_api.c [hunk 1]
```diff
@@ -0,0 +1,390 @@
...
+
+	ASSERT_RESULT(res);
+
+	wrapped_handle = allocate_wrapped_feature_handle(
+		wrapped_token, feature_handle, wrapped_token->adapter_table);
+
+	if (!wrapped_handle) {
+		FPGA_ERR("malloc failed");
+		res = FPGA_NO_MEMORY;
+		cres = wrapped_token->adapter_table->fpgaFeatureClose(feature_handle);
```
Comment (tswhison) [inline libopae/plugins/xfpga/feature_plugins/feature_api.c]: You haven't verified that fpgaFeatureClose is non-NULL before calling it.

File: libopae/plugins/xfpga/feature_plugins/feature_api.c [hunk 2]
```diff
@@ -0,0 +1,390 @@
...
+
+	return res;
+}	
+	
+fpga_result __FPGA_API__
+xfpga_fpgaDMAPropertiesGet(fpga_feature_token token, fpgaDMAProperties *prop,
+									int max_ch)
+{
+	wrapped_feature_token *wrapped_token =
+				validate_wrapped_feature_token(token);;
```
Comment (tswhison) [inline libopae/plugins/xfpga/feature_plugins/feature_api.c]: s/;;/;/

File: libopae/plugins/xfpga/feature_plugins/feature_int.h [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+
+#define UNUSED_PARAM(x) ((void)x)
+
+typedef struct _feature_adapter_table feature_adapter_table;
+//                                  f e a t    
+#define FPGA_WRAPPED_FEATURE_TOKEN_MAGIC 0x66656174
+#define FPGA_FEATURE_TOKEN_MAGIC 0x66656178
+
+/** Device-wide unique FPGA feature resource identifier */
+struct _fpga_feature_token {
```
Comment (tswhison) [inline libopae/plugins/xfpga/feature_plugins/feature_int.h]: There is already a type named fpga_feature_token.  Can this be called something else for clarity?

File: libopae/plugins/xfpga/feature_plugins/feature_int.h [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+		return NULL;
+	wt = (wrapped_feature_token *)t;
+	return (wt->magic == FPGA_WRAPPED_FEATURE_TOKEN_MAGIC) ? wt : NULL;
+}
+
+static inline void destroy_wrapped_feature_token(wrapped_feature_token *wt)
+{
+	wt->magic = 0;
+	free(wt);
+	wt = NULL;
```
Comment (tswhison) [inline libopae/plugins/xfpga/feature_plugins/feature_int.h]: Statement has no effect.  ie, this just sets parameter "wt" to NULL.

File: libopae/plugins/xfpga/feature_plugins/feature_int.h [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+		return NULL;
+	wh = (wrapped_feature_handle *)h;
+	return (wh->magic == FPGA_WRAPPED_FEATURE_HANDLE_MAGIC) ? wh : NULL;
+}
+
+static inline void destroy_wrapped_feature_handle(wrapped_feature_handle *wh)
+{
+	wh->magic = 0;
+	free(wh);
+	wh = NULL;
```
Comment (tswhison) [inline libopae/plugins/xfpga/feature_plugins/feature_int.h]: Statement has no effect. ie, this just sets parameter "wh" to NULL.

File: libopae/plugins/xfpga/feature_plugins/feature_pluginmgr.c [hunk 6]
```diff
@@ -0,0 +1,475 @@
...
+			OPAE_DBG("platform detected: feature_id=%s -> %s",
+				 feature_id feature_data_table[i]
+					 .feature_plugin);
+
+			feature_data_table[i].flags |=
+				FPGA_FEATURE_DATA_DETECTED;
+		}
+	}
+
+out_close:
```
Comment (tswhison) [inline libopae/plugins/xfpga/feature_plugins/feature_pluginmgr.c]: out_close?  Nothing is being closed.

File: libopae/plugins/xfpga/feature_plugins/feature_pluginmgr.c [hunk 7]
```diff
@@ -0,0 +1,475 @@
...
+			       strnlen_s(feature_data_table[i].feature_id, 256),
+			       feature_id, &res)) {
+			FPGA_ERR("strcmp_s failed");
+			++errors;
+			goto out_close;
+		}
+
+		if (0 == res) {
+			OPAE_DBG("platform detected: feature_id=%s -> %s",
+				 feature_id feature_data_table[i]
```
Comment (tswhison) [inline libopae/plugins/xfpga/feature_plugins/feature_pluginmgr.c]: Needs a comma after feature_id.

File: libopae/plugins/xfpga/feature_plugins/feature_pluginmgr.c [hunk 8]
```diff
@@ -0,0 +1,475 @@
...
+
+	for (i = 14; i >= 0; i -= 2) {
+		p = id_hi + i;
+		sscanf_s_u(p, "%2x", &tmp);
+		(*guid)[15 - i / 2] = tmp;
+		p[0] = '\0';
+	}
+}
+
+static fpga_result opae_plugin_mgr_detect_features(fpga_handle handle)
```
Comment (tswhison) [inline libopae/plugins/xfpga/feature_plugins/feature_pluginmgr.c]: Do we know that handle is locked before this call?

File: libopae/plugins/xfpga/feature_plugins/feature_pluginmgr.c [hunk 9]
```diff
@@ -0,0 +1,475 @@
...
+			return res;
+		}
+
+		get_guid(feature_uuid_lo, feature_uuid_hi, &guid);
+		
+		// Detect feature for this guid.
+		opae_plugin_mgr_detect_feature(guid);
+
+		// Move to the next feature header
+		offset = offset + dfh.next_header_offset;
```
Comment (tswhison) [inline libopae/plugins/xfpga/feature_plugins/feature_pluginmgr.c]: Need to update dfh here.

File: libopae/plugins/xfpga/feature_plugins/feature_token_list.c [hunk 10]
```diff
@@ -0,0 +1,161 @@
+// Copyright(c) 2017-2018, Intel Corporation
```
Comment (tswhison) [inline libopae/plugins/xfpga/feature_plugins/feature_token_list.c]: s/2017-2018/2018/

File: libopae/plugins/xfpga/feature_plugins/feature_token_list.c [hunk 11]
```diff
@@ -0,0 +1,161 @@
...
+#undef _GNU_SOURCE
+
+#include "safe_string/safe_string.h"
+#include "types_int.h"
+#include "feature_token_list_int.h"
+
+/* global list of tokens we've seen */
+static struct _fpga_feature_token *ftoken_root;
+/** Mutex to protect feature tokens */
+pthread_mutex_t ftoken_lock = PTHREAD_MUTEX_INITIALIZER;
```
Comment (tswhison) [inline libopae/plugins/xfpga/feature_plugins/feature_token_list.c]: Probably want a recursive mutex here.

File: libopae/plugins/xfpga/feature_plugins/feature_token_list.c [hunk 12]
```diff
@@ -0,0 +1,161 @@
...
+			}
+			return tmp;
+		}
+	}
+
+	tmp = (struct _fpga_feature_token *)malloc(
+		sizeof(struct _fpga_feature_token));
+	if (NULL == tmp) {
+		FPGA_ERR("Failed to allocate memory for fhandle");
+		return NULL;
```
Comment (tswhison) [inline libopae/plugins/xfpga/feature_plugins/feature_token_list.c]: ftoken_lock is still locked.

File: libopae/plugins/xfpga/feature_plugins/feature_token_list.c [hunk 13]
```diff
@@ -0,0 +1,161 @@
...
+		goto out_unlock;
+
+	while (current) {
+		struct _fpga_feature_token *tmp = current;
+		current = current->next;
+
+		// invalidate magic (just in case)
+		tmp->magic = FPGA_INVALID_MAGIC;
+		free(tmp);
+		tmp = NULL;
```
Comment (tswhison) [inline libopae/plugins/xfpga/feature_plugins/feature_token_list.c]: Statement has no effect.

File: libopae/plugins/xfpga/feature_plugins/feature_token_list_int.h [hunk 14]
```diff
@@ -0,0 +1,39 @@
+// Copyright(c) 2017, Intel Corporation
```
Comment (tswhison) [inline libopae/plugins/xfpga/feature_plugins/feature_token_list_int.h]: s/2017/2018/

File: testing/CMakeLists.txt [hunk 15]
```diff
@@ -72,13 +72,17 @@ add_library(xfpga-static STATIC
               ${OPAE_SDK_SOURCE}/libopae/plugins/xfpga/metrics/metrics_utils.c
               ${OPAE_SDK_SOURCE}/libopae/plugins/xfpga/metrics/afu_metrics.c
               ${OPAE_SDK_SOURCE}/libopae/plugins/xfpga/metrics/metrics.c
-              ${OPAE_SDK_SOURCE}/libopae/plugins/xfpga/metrics/vector.c)
+              ${OPAE_SDK_SOURCE}/libopae/plugins/xfpga/metrics/vector.c
+              ${OPAE_SDK_SOURCE}/libopae/plugins/xfpga/feature_plugins/feature_pluginmgr.c
+              ${OPAE_SDK_SOURCE}/libopae/plugins/xfpga/feature_plugins/feature_token_list.c
+			  ${OPAE_SDK_SOURCE}/libopae/plugins/xfpga/feature_plugins/feature_api.c)
```
Comment (tswhison) [inline testing/CMakeLists.txt]: Alignment is off.

File: testing/xfpga/test_feature_enum_c.cpp [hunk 16]
```diff
@@ -0,0 +1,226 @@
+// Copyright(c) 2017-2018, Intel Corporation
```
Comment (tswhison) [inline testing/xfpga/test_feature_enum_c.cpp]: s/2017-2018/2018/

File: testing/xfpga/test_feature_open_c.cpp [hunk 17]
```diff
@@ -0,0 +1,281 @@
+// Copyright(c) 2017-2018, Intel Corporation
```
Comment (tswhison) [inline testing/xfpga/test_feature_open_c.cpp]: s/2017-2018/2018/

File: testing/xfpga/test_feature_token_list_c.cpp [hunk 18]
```diff
@@ -0,0 +1,73 @@
+
```
Comment (tswhison) [inline testing/xfpga/test_feature_token_list_c.cpp]: Remove blank line.

File: testing/xfpga/test_feature_token_list_c.cpp [hunk 19]
```diff
@@ -0,0 +1,73 @@
...
+// Copyright(c) 2017-2018, Intel Corporation
```
Comment (tswhison) [inline testing/xfpga/test_feature_token_list_c.cpp]: s/2017-2018/2018/

File: common/include/opae/types_enum.h [hunk 20]
```diff
@@ -177,4 +177,33 @@ enum fpga_metric_datatype {
...
+/**
+ * fpga dma transfer type
+ *
+ * Indicate the required transfer type.
+ * For streaming: host to AFU, AFU to host, FPGA local mem to AFU, AFU to FPGA
+ * local mem memory to memory: host to fpga, fpga to host, fpga internal  memory
+ * copy
+ */
+typedef enum {
+	HOST_MM_TO_FPGA_ST = (1u << 0), /**< streaming, host to AFU */
```
[the remark this answers is not in the document]
Comment (lzhan55) [reply common/include/opae/types_enum.h]: MM-> Memory to Memory transfer
ST -> Streaming DMA

File: libopae/api-shell.c [hunk 21]
```diff
@@ -1540,3 +1573,202 @@ fpga_result fpgaGetMetricsByName(fpga_handle handle,
...
+		if (!d_tokens) {
+			OPAE_ERR("out of memory");
+			return FPGA_NO_MEMORY;
+		}
+	}
+	res = wrapped_handle->adapter_table->fpgaFeatureEnumerate(wrapped_handle->opae_handle,
+							prop, d_tokens,	max_tokens, num_matches);
+	if (res != FPGA_OK) {
+		OPAE_ERR("fpgaFeatureEnumerate failed");
+		return res;
```
[the remark this answers is not in the document]
Comment (lzhan55) [reply libopae/api-shell.c]: good catch!

File: libopae/api-shell.c [hunk 22]
```diff
@@ -1540,3 +1573,202 @@ fpga_result fpgaGetMetricsByName(fpga_handle handle,
...
+		num_tokens = max_tokens;
+	else
+		num_tokens = *num_matches;
+
+	for (i = 0; i < num_tokens; ++i) {
+		opae_wrapped_feature_token *wt = opae_allocate_wrapped_feature_token(
+			d_tokens[i], wrapped_handle->adapter_table);
+		if (!wt) {
+			OPAE_ERR("opae_allocate_wrapped_feature_token failed");
+			return OPAE_ENUM_STOP;
```
[the remark this answers is not in the document]
Comment (lzhan55) [reply libopae/api-shell.c]: will check and fix!

File: libopae/api-shell.c [hunk 23]
```diff
@@ -1540,3 +1573,202 @@ fpga_result fpgaGetMetricsByName(fpga_handle handle,
...
+	opae_destroy_wrapped_feature_handle(wrapped_handle);
+
+	return res;
+}	
+	
+fpga_result fpgaDMAPropertiesGet(fpga_feature_token token, fpgaDMAProperties *prop,
+									int max_ch)
+{
+	opae_wrapped_feature_token *wrapped_token =
+				opae_validate_wrapped_feature_token(token);;
```
[the remark this answers is not in the document]
Comment (lzhan55) [reply libopae/api-shell.c]: Good catch!

File: libopae/opae_int.h [hunk 24]
```diff
@@ -219,4 +219,68 @@ static inline void opae_destroy_wrapped_object(opae_wrapped_object *wo)
...
+		return NULL;
+	wt = (opae_wrapped_feature_token *)t;
+	return (wt->magic == OPAE_WRAPPED_FEATURE_TOKEN_MAGIC) ? wt : NULL;
+}
+
+static inline void opae_destroy_wrapped_feature_token(opae_wrapped_feature_token *wt)
+{
+	wt->magic = 0;
+	free(wt);
+	wt = NULL;
```
[the remark this answers is not in the document]
Comment (lzhan55) [reply libopae/opae_int.h]: Isn't this the SDL requirement to set the pointer to null after freeing?

File: libopae/plugins/xfpga/feature_plugins/dma_sav/CMakeLists.txt [hunk 25]
```diff
@@ -0,0 +1,6 @@
+cmake_minimum_required (VERSION 2.8)
```
[the remark this answers is not in the document]
Comment (lzhan55) [reply libopae/plugins/xfpga/feature_plugins/dma_sav/CMakeLists.txt]: This is the stub library which will be removed when the real DMA plugin library is ready.

File: libopae/plugins/xfpga/feature_plugins/dma_sav/intel-dma/CMakeLists.txt [hunk 26]
```diff
@@ -0,0 +1,122 @@
...
+## ARE DISCLAIMED.  IN NO EVENT  SHALL THE COPYRIGHT OWNER  OR CONTRIBUTORS BE
+## LIABLE  FOR  ANY  DIRECT,  INDIRECT,  INCIDENTAL,  SPECIAL,  EXEMPLARY,  OR
+## CONSEQUENTIAL  DAMAGES  (INCLUDING,  BUT  NOT LIMITED  TO,  PROCUREMENT  OF
+## SUBSTITUTE GOODS OR SERVICES;  LOSS OF USE,  DATA, OR PROFITS;  OR BUSINESS
+## INTERRUPTION)  HOWEVER CAUSED  AND ON ANY THEORY  OF LIABILITY,  WHETHER IN
+## CONTRACT,  STRICT LIABILITY,  OR TORT  (INCLUDING NEGLIGENCE  OR OTHERWISE)
+## ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE,  EVEN IF ADVISED OF THE
+## POSSIBILITY OF SUCH DAMAGE.
+
+#project(libxfpga)
```
[the remark this answers is not in the document]
Comment (lzhan55) [reply libopae/plugins/xfpga/feature_plugins/dma_sav/intel-dma/CMakeLists.txt]: This is the throw-away stub library.

File: libopae/plugins/xfpga/feature_plugins/feature_api.c [hunk 27]
```diff
@@ -0,0 +1,390 @@
...
+		whan->magic = FPGA_WRAPPED_FEATURE_HANDLE_MAGIC;
+		whan->wrapped_feature_token = wt;
+		whan->feature_handle = feature_handle;
+		whan->adapter_table = adapter;
+	}
+
+	return whan;
+}
+
+static fpga_result
```
[the remark this answers is not in the document]
Comment (lzhan55) [reply libopae/plugins/xfpga/feature_plugins/feature_api.c]: Will remove the static keyword.

File: libopae/plugins/xfpga/feature_plugins/feature_api.c [hunk 28]
```diff
@@ -0,0 +1,390 @@
...
+  
+	*num_matches = 0;
+	
+	errors = feature_plugin_mgr_initialize(handle);
+	if (errors) {
+		FPGA_ERR("Feature token initialize errors");
+		result = FPGA_EXCEPTION;
+	}	
+	
+	/*res = fpgaGetProperties(wrapped_handle->wrapped_token, &accel_prop);
```
[the remark this answers is not in the document]
Comment (lzhan55) [reply libopae/plugins/xfpga/feature_plugins/feature_api.c]: sure

File: libopae/plugins/xfpga/feature_plugins/feature_api.c [hunk 29]
```diff
@@ -0,0 +1,390 @@
...
+	res = fpgaPropertiesGetNumMMIO(accel_prop, &mmio_num);
+	if (res != FPGA_OK) {
+		FPGA_ERR("fpgaPropertiesGetNumMMIO() failed");
+		return res;
+	} */
+	mmio_num = 0;  // TODO : debug above
+	result = xfpga_fpgaReadMMIO64(handle, mmio_num, 0x0, &(dfh.csr));
+	if (result != FPGA_OK) {
+		FPGA_ERR("fpgaReadMMIO64() failed");
+		return result;
```
[the remark this answers is not in the document]
Comment (lzhan55) [reply libopae/plugins/xfpga/feature_plugins/feature_api.c]: Will check the locking mechanism. Thanks.

File: libopae/plugins/xfpga/feature_plugins/feature_api.c (hunk 1 above)
Comment (lzhan55) [reply libopae/plugins/xfpga/feature_plugins/feature_api.c]: Will fix

File: libopae/plugins/xfpga/feature_plugins/feature_int.h (hunk 3 above)
Comment (lzhan55) [reply libopae/plugins/xfpga/feature_plugins/feature_int.h]: Will figure out a new name to make it clear.

**Comment by lzhan55** [comment]
@tswhison Thank you for your suggestions about the locking, naming convention and a lot of other detailed suggestions! I appreciate your early feedbacks!

File: libopae/api-shell.c [hunk 30]
```diff
@@ -1540,3 +1573,202 @@ fpga_result fpgaGetMetricsByName(fpga_handle handle,
...
+		num_tokens = *num_matches;
+
+	for (i = 0; i < num_tokens; ++i) {
+		opae_wrapped_feature_token *wt = opae_allocate_wrapped_feature_token(
+			d_tokens[i], wrapped_handle->adapter_table);
+		if (!wt) {
+			OPAE_ERR("opae_allocate_wrapped_feature_token failed");
+			return OPAE_ENUM_STOP;
+		}
+		tokens[i] = wt;
```
[the remark this answers is not in the document]
Comment (lzhan55) [reply libopae/api-shell.c]: Will add code to check tokens not NULL.

File: libopae/plugins/xfpga/feature_plugins/feature_api.c [hunk 31]
```diff
@@ -0,0 +1,390 @@
...
+					feature_adapter_table * adapter;
+					
+					result = xfpga_fpgaCloneFeatureToken((fpga_feature_token)_ftoken, &tmp);
+					if (result	!= FPGA_OK) {
+						FPGA_MSG("Error cloning token");
+					}
+					adapter = get_feature_plugin_adapter(guid);
+					wrapped_feature_token *wt = 
+					     allocate_wrapped_feature_token(tmp, adapter);
+					tokens[*num_matches] = wt;
```
[the remark this answers is not in the document]
Comment (lzhan55) [reply libopae/plugins/xfpga/feature_plugins/feature_api.c]: Will allow tokens to be NULL.

File: libopae/plugins/xfpga/feature_plugins/feature_api.c [hunk 32]
```diff
@@ -0,0 +1,390 @@
...
+				}
+				++(*num_matches);
+			}
+		}
+
+		// Move to the next feature header 
+		offset = offset + dfh.next_header_offset;
```
[the remark this answers is not in the document]
Comment (lzhan55) [reply libopae/plugins/xfpga/feature_plugins/feature_api.c]: dfh.next_header_offset should move to the next header's MMIO offset, right?

[the remark this answers is not in the document]
Comment (tswhison) [reply libopae/plugins/xfpga/feature_plugins/feature_api.c]: The loop condition is
```c
} while (!dfh.eol);
```
but I don't see that dfh is being updated.  Did I miss it?

**Pull request closed without merging**